m2cgen/interpreters/code_generator.py

```diff
@@ -1,17 +1,18 @@
-from string import Template
-
-
 class CodeTemplate:
 
     def __init__(self, template):
-        self.template = Template(template)
         self.str_template = template
 
     def __str__(self):
         return self.str_template
 
     def __call__(self, *args, **kwargs):
-        return self.template.substitute(*args, **kwargs)
+        # Force calling str() representation
+        # because without it numpy gives the same output
+        # for different float types
+        return self.str_template.format(
+            *[str(i) for i in args],
+            **{k: str(v) for k, v in kwargs.items()})
```

Comment:
This bug(?) was spotted by our XGBoost(hist) e2e test and can be demonstrated in the following example.

```
import numpy as np

test_float_val = 42.42424242424242424242

print(np.float16(test_float_val).__str__())
# 42.44  Correct
print(np.float32(test_float_val).__str__())
# 42.424244  Correct
print(np.float64(test_float_val).__str__())
# 42.42424242424242  Correct

print(np.float16(test_float_val).__format__(""))
# 42.4375  Wrong
print(np.float32(test_float_val).__format__(""))
# 42.42424392700195  Wrong
print(np.float64(test_float_val).__format__(""))
# 42.42424242424242  Correct
```

Reply:
Oh wow, this is a great catch. Does it mean that the root cause we originally discovered could've been wrong?

Reply:
Sorry, didn't get it!

This issue illustrates only that we should always dump float values into native code via `str()` and not `format()`. 

Our current code generator use `template.substitute()` which calls `str()` under the hood. So we wasn't able to spot this "bug" (I'm still not sure that this is a real bug, or `numpy` produces different results with `str()` and `format()` by design).